File: js/FileSystem/FileSystemTree.js
import { unwrap, defaultArg } from "../fable_modules/fable-library-js.4.22.0/Option.js";
import { append as append_1, choose, equalsWith, tail, head, map, tryFind } from "../fable_modules/fable-library-js.4.22.0/Array.js";
import { singleton, append, delay, toArray } from "../fable_modules/fable-library-js.4.22.0/Seq.js";
import { Array_distinct, Array_groupBy } from "../fable_modules/fable-library-js.4.22.0/Seq2.js";
import { curry2, arrayHash, stringHash } from "../fable_modules/fable-library-js.4.22.0/Util.js";
import { combineMany, split } from "./Path.js";
import { empty, reverse, toArray as toArray_1, cons } from "../fable_modules/fable-library-js.4.22.0/List.js";
import { Union } from "../fable_modules/fable-library-js.4.22.0/Types.js";
import { union_type, array_type, string_type } from "../fable_modules/fable-library-js.4.22.0/Reflection.js";

export class FileSystemTree extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["File", "Folder"];
    }
    get Name() {
        const this$ = this;
        let n;
        if (this$.tag === 1) {
            n = this$.fields[0];
        }
        else {
            n = this$.fields[0];
        }
        return n;
    }
    get isFolder() {
        const this$ = this;
        return this$.tag === 1;
    }
    get isFile() {
        const this$ = this;
        return this$.tag === 0;
    }
    static get ROOT_NAME() {
        return "root";
    }
    static createFile(name) {
        return new FileSystemTree(0, [name]);
    }
    static createFolder(name, children) {
        return new FileSystemTree(1, [name, defaultArg(children, [])]);
    }
    static createRootFolder(children) {
        return new FileSystemTree(1, [FileSystemTree.ROOT_NAME, children]);
    }
    TryGetChildByName(name) {
        const this$ = this;
        return (this$.tag === 0) ? undefined : tryFind((c) => (c.Name === name), this$.fields[1]);
    }
    static tryGetChildByName(name) {
        return (fst) => fst.TryGetChildByName(name);
    }
    ContainsChildWithName(name) {
        const this$ = this;
        return (this$.tag === 0) ? false : this$.fields[1].some((c) => (c.Name === name));
    }
    static containsChildWithName(name) {
        return (fst) => fst.ContainsChildWithName(name);
    }
    AddFile(path) {
        const this$ = this;
        const existingPaths = this$.ToFilePaths();
        const filePaths = toArray(delay(() => append(singleton(path), delay(() => existingPaths))));
        return FileSystemTree.fromFilePaths(filePaths);
    }
    static addFile(path) {
        return (tree) => tree.AddFile(path);
    }
    static fromFilePaths(paths) {
        const loop = (paths_1, parent) => {
            const files = map((arg) => {
                const name = head(arg);
                return FileSystemTree.createFile(name);
            }, paths_1.filter((p) => (p.length === 1)));
            const folders = map((tupledArg) => {
                const parent_1 = FileSystemTree.createFolder(tupledArg[0], []);
                return loop(map(tail, tupledArg[1]), parent_1);
            }, Array_groupBy(head, paths_1.filter((p_1) => (p_1.length > 1)), {
                Equals: (x_3, y_2) => (x_3 === y_2),
                GetHashCode: stringHash,
            }));
            if (parent.tag === 1) {
                return FileSystemTree.createFolder(parent.fields[0], toArray(delay(() => append(files, delay(() => folders)))));
            }
            else {
                return parent;
            }
        };
        return loop(Array_distinct(map(split, paths), {
            Equals: (x, y) => equalsWith((x_1, y_1) => (x_1 === y_1), x, y),
            GetHashCode: arrayHash,
        }), FileSystemTree.createFolder(FileSystemTree.ROOT_NAME, []));
    }
    ToFilePaths(removeRoot) {
        const this$ = this;
        const res = [];
        const loop = (output, parent) => {
            if (parent.tag === 1) {
                parent.fields[1].forEach((filest) => {
                    loop(cons(parent.fields[0], output), filest);
                });
            }
            else {
                const item = combineMany(toArray_1(reverse(cons(parent.fields[0], output))));
                void (res.push(item));
            }
        };
        if (defaultArg(removeRoot, true)) {
            if (this$.tag === 0) {
                void (res.push(this$.fields[0]));
            }
            else {
                const action_1 = curry2(loop)(empty());
                this$.fields[1].forEach(action_1);
            }
        }
        else {
            loop(empty(), this$);
        }
        return Array.from(res);
    }
    static toFilePaths(removeRoot) {
        return (root) => root.ToFilePaths(unwrap(removeRoot));
    }
    FilterFiles(predicate) {
        const this$ = this;
        const loop = (parent) => {
            if (parent.tag === 1) {
                return new FileSystemTree(1, [parent.fields[0], choose(loop, parent.fields[1])]);
            }
            else {
                const n = parent.fields[0];
                if (predicate(n)) {
                    return new FileSystemTree(0, [n]);
                }
                else {
                    return undefined;
                }
            }
        };
        return loop(this$);
    }
    static filterFiles(predicate) {
        return (tree) => tree.FilterFiles(predicate);
    }
    FilterFolders(predicate) {
        const this$ = this;
        const loop = (parent) => {
            if (parent.tag === 1) {
                const n_1 = parent.fields[0];
                if (predicate(n_1)) {
                    return new FileSystemTree(1, [n_1, choose(loop, parent.fields[1])]);
                }
                else {
                    return undefined;
                }
            }
            else {
                return new FileSystemTree(0, [parent.fields[0]]);
            }
        };
        return loop(this$);
    }
    static filterFolders(predicate) {
        return (tree) => tree.FilterFolders(predicate);
    }
    Filter(predicate) {
        const this$ = this;
        const loop = (parent) => {
            if (parent.tag === 1) {
                const n_1 = parent.fields[0];
                if (predicate(n_1)) {
                    const filteredChildren = choose(loop, parent.fields[1]);
                    if (filteredChildren.length === 0) {
                        return undefined;
                    }
                    else {
                        return new FileSystemTree(1, [n_1, filteredChildren]);
                    }
                }
                else {
                    return undefined;
                }
            }
            else {
                const n = parent.fields[0];
                if (predicate(n)) {
                    return new FileSystemTree(0, [n]);
                }
                else {
                    return undefined;
                }
            }
        };
        return loop(this$);
    }
    static filter(predicate) {
        return (tree) => tree.Filter(predicate);
    }
    Union(otherFST) {
        let array_1;
        const this$ = this;
        const paths = Array_distinct((array_1 = this$.ToFilePaths(), append_1(otherFST.ToFilePaths(), array_1)), {
            Equals: (x, y) => (x === y),
            GetHashCode: stringHash,
        });
        return FileSystemTree.fromFilePaths(paths);
    }
    Copy() {
        const this$ = this;
        return (this$.tag === 0) ? (new FileSystemTree(0, [this$.fields[0]])) : (new FileSystemTree(1, [this$.fields[0], map((c) => c.Copy(), this$.fields[1])]));
    }
    static createGitKeepFile() {
        return FileSystemTree.createFile(".gitkeep");
    }
    static createReadmeFile() {
        return FileSystemTree.createFile("README.md");
    }
    static createEmptyFolder(name) {
        return FileSystemTree.createFolder(name, [FileSystemTree.createGitKeepFile()]);
    }
    static createAssayFolder(assayName, hasDataMap) {
        const hasDataMap_1 = defaultArg(hasDataMap, false);
        const dataset = FileSystemTree.createEmptyFolder("dataset");
        const protocols = FileSystemTree.createEmptyFolder("protocols");
        const readme = FileSystemTree.createReadmeFile();
        const assayFile = FileSystemTree.createFile("isa.assay.xlsx");
        if (hasDataMap_1) {
            const dataMapFile = FileSystemTree.createFile("isa.datamap.xlsx");
            return FileSystemTree.createFolder(assayName, [dataset, protocols, assayFile, readme, dataMapFile]);
        }
        else {
            return FileSystemTree.createFolder(assayName, [dataset, protocols, assayFile, readme]);
        }
    }
    static createStudyFolder(studyName, hasDataMap) {
        const hasDataMap_1 = defaultArg(hasDataMap, false);
        const resources = FileSystemTree.createEmptyFolder("resources");
        const protocols = FileSystemTree.createEmptyFolder("protocols");
        const readme = FileSystemTree.createReadmeFile();
        const studyFile = FileSystemTree.createFile("isa.study.xlsx");
        if (hasDataMap_1) {
            const dataMapFile = FileSystemTree.createFile("isa.datamap.xlsx");
            return FileSystemTree.createFolder(studyName, [resources, protocols, studyFile, readme, dataMapFile]);
        }
        else {
            return FileSystemTree.createFolder(studyName, [resources, protocols, studyFile, readme]);
        }
    }
    static createInvestigationFile() {
        return FileSystemTree.createFile("isa.investigation.xlsx");
    }
    static createAssaysFolder(assays) {
        return FileSystemTree.createFolder("assays", append_1([FileSystemTree.createGitKeepFile()], assays));
    }
    static createStudiesFolder(studies) {
        return FileSystemTree.createFolder("studies", append_1([FileSystemTree.createGitKeepFile()], studies));
    }
    static createWorkflowsFolder(workflows) {
        return FileSystemTree.createFolder("workflows", append_1([FileSystemTree.createGitKeepFile()], workflows));
    }
    static createRunsFolder(runs) {
        return FileSystemTree.createFolder("runs", append_1([FileSystemTree.createGitKeepFile()], runs));
    }
}

export function FileSystemTree_$reflection() {
    return union_type("ARCtrl.FileSystem.FileSystemTree", [], FileSystemTree, () => [[["name", string_type]], [["name", string_type], ["children", array_type(FileSystemTree_$reflection())]]]);
}

File: src/FileSystem/FileSystemTree.fs
namespace ARCtrl.FileSystem

open Fable.Core

open ARCtrl

[<AttachMembers>]
type FileSystemTree =
    | File of name:string
    | Folder of name:string * children:FileSystemTree array

    member this.Name = match this with | File n | Folder (n,_) -> n
    member this.isFolder = match this with | Folder _ -> true | _ -> false
    member this.isFile = match this with | File _ -> true | _ -> false
    /// <summary>
    /// Defines the name of the root folder of the FileSystemTree when created by `fromFilePaths`.
    /// </summary>
    static member ROOT_NAME = "root"

    static member createFile name = File name

    static member createFolder(name:string, ?children:FileSystemTree array) = 
        let children = defaultArg children [||]
        Folder(name,children)

    static member createRootFolder(children:FileSystemTree array) = 
        Folder(FileSystemTree.ROOT_NAME,children)
        
    /// Non-recursive lookup of child with the given name
    member this.TryGetChildByName(name : string) = 
        match this with
        | Folder (_,children) ->
            children
            |> Array.tryFind (fun c -> c.Name = name)
        | File _ -> None

    static member tryGetChildByName (name : string) =
        fun (fst : FileSystemTree) -> fst.TryGetChildByName name

    /// Non-recursive lookup of child with the given name
    member this.ContainsChildWithName(name : string) =
        match this with
        | Folder (_,children) ->
            children
            |> Array.exists (fun c -> c.Name = name)
        | File _ -> false

    /// Non-recursive lookup of child with the given name
    static member containsChildWithName (name : string) =
        fun (fst : FileSystemTree) -> fst.ContainsChildWithName name


    member this.TryGetPath(path:string) : FileSystemTree option =
        let rec loop (parent: FileSystemTree) (pathParts: string array) =
            match parent with
            | Folder ("root", children) ->
                children |> Array.tryPick (fun child -> loop child pathParts)
            | Folder (name, children) ->
                let nextPart = Array.head pathParts
                if name = nextPart then
                    if pathParts.Length = 1 then
                        Some parent
                    else
                        let remainingParts = Array.tail pathParts
                        children |> Array.tryPick (fun child -> loop child remainingParts)
                else
                    None // Current folder name does not match the next part of the path
            | File name -> 
                if pathParts.Length = 1 && pathParts.[0] = name then
                    Some parent
                else 
                    None // Cannot go deeper into a file
        let parts = ARCtrl.ArcPathHelper.split path
        loop this parts

    member this.AddFile (path: string) : FileSystemTree =
        let existingPaths = this.ToFilePaths()
        let filePaths = [|
            path
            yield! existingPaths
        |]
        FileSystemTree.fromFilePaths (filePaths)

    static member addFile (path: string) =
        fun (tree: FileSystemTree) -> tree.AddFile(path)

    /// <summary>
    /// Split `paths` by `\\` or `/` and sort them into FileSystemTree. All given paths must be relative to ARC root and are sorted into a folder called `root`.
    /// </summary>
    /// <param name="paths">A array of file paths relative to ARC root.</param>
    static member fromFilePaths (paths: string array) : FileSystemTree =
        // Split path by seperators into path sequence.
        let splitPaths = paths |> Array.map ARCtrl.ArcPathHelper.split |> Array.distinct
        let root = FileSystemTree.createFolder(FileSystemTree.ROOT_NAME,[||])
        let rec loop (paths:string [] []) (parent: FileSystemTree) =
            // Files are always the last and only in path sequence.
            let files = paths |> Array.filter (fun p -> p.Length = 1) |> Array.map (Array.head >> FileSystemTree.createFile)
            // If we have more than one element in path sequence the first one must still be a folder.
            let folders = 
                paths 
                |> Array.filter (fun p -> p.Length > 1) 
                |> Array.groupBy(fun x -> Array.head x)
                |> Array.map (fun (folderName, children) -> 
                    let parent = FileSystemTree.createFolder(folderName, [||])
                    let children = children |> Array.map Array.tail
                    loop children parent 
                )
            match parent with
            | Folder (name, _) -> FileSystemTree.createFolder(name, [|yield! files; yield! folders|])
            | x -> x
        loop splitPaths root 

    /// <summary>
    /// Reverts FileSystemTree back to an array of filepaths relative to ARC root.
    /// </summary>
    /// <param name="removeRoot">Will remove root `Folder` if set true. *Default*: `true`.</param>
    member this.ToFilePaths (?removeRoot: bool) =
        let removeRoot = defaultArg removeRoot true
        let res = ResizeArray<string>()
        let rec loop (output: string list) (parent: FileSystemTree)  =
            match parent with
            | File n -> 
                (n::output)
                |> List.rev 
                |> Array.ofList 
                |> ARCtrl.ArcPathHelper.combineMany 
                |> res.Add //output full path
            | Folder (n, children) ->
                let nextOutput = n::output
                children
                |> Array.iter (fun filest -> 
                    loop nextOutput filest 
                )
        // When creating FileSystemTree with `fromFilePaths` we create an `root` `Folder`. 
        // This root folder serves as relative parent for all filepaths given.
        // The removeRoot flag will remove this extra folder.
        if removeRoot then
            match this with
            | Folder (_, children) -> 
                children |> Array.iter (loop [])
            | File n -> res.Add(n)
        else
            loop [] this
        res
        |> Array.ofSeq

    static member toFilePaths (?removeRoot: bool) =
        fun (root: FileSystemTree) -> root.ToFilePaths(?removeRoot=removeRoot)

    member this.FilterFiles (predicate: string -> bool) =
        let rec loop (parent: FileSystemTree) =
            match parent with
            | File n -> 
                if predicate n then Some (File n) else None
            | Folder (n, children) ->
                Folder (n, children |> Array.choose loop)
                |> Some

        loop this

    static member filterFiles (predicate: string -> bool) =
        fun (tree: FileSystemTree) -> tree.FilterFiles predicate

    member this.FilterFolders (predicate: string -> bool) =
        let rec loop (parent: FileSystemTree) =
            match parent with
            | File n -> Some (File n)
            | Folder (n, children) ->
                if predicate n then 
                    Folder (n, children |> Array.choose loop)
                    |> Some
                else
                    None
        loop this

    static member filterFolders (predicate: string -> bool) =
        fun (tree: FileSystemTree) -> tree.FilterFolders predicate

    member this.Filter (predicate: string -> bool) = 
        let rec loop (parent: FileSystemTree) = 
            match parent with 
            | File n ->  
                if predicate n then Some (FileSystemTree.File n) else None 
            | Folder (n, children) -> 
                if predicate n then 
                    let filteredChildren = children |> Array.choose loop 
                    if Array.isEmpty filteredChildren then  
                        None
                    else
                        Some (FileSystemTree.Folder (n,filteredChildren))
                else
                    None
        loop this 

    static member filter (predicate: string -> bool) =
        fun (tree: FileSystemTree) -> tree.Filter predicate

    member this.Union(otherFST : FileSystemTree) =
        this.ToFilePaths() 
        |> Array.append (otherFST.ToFilePaths())
        |> Array.distinct
        |> FileSystemTree.fromFilePaths

    member this.Copy() =
        match this with
        | Folder(name,children) -> 
            Folder (name, children |> Array.map (fun c -> c.Copy()))
        | File(name) -> 
            File name




    static member createGitKeepFile() = 
        FileSystemTree.createFile ARCtrl.ArcPathHelper.GitKeepFileName

    static member createReadmeFile() = 
        FileSystemTree.createFile ARCtrl.ArcPathHelper.READMEFileName

    static member createEmptyFolder (name : string) = 
        FileSystemTree.createFolder(name, [|FileSystemTree.createGitKeepFile()|])

    static member createAssayFolder(assayName : string, ?hasDataMap) = 
        let hasDataMap = defaultArg hasDataMap false
        let dataset = FileSystemTree.createEmptyFolder ARCtrl.ArcPathHelper.AssayDatasetFolderName
        let protocols = FileSystemTree.createEmptyFolder ARCtrl.ArcPathHelper.AssayProtocolsFolderName
        let readme = FileSystemTree.createReadmeFile()
        let assayFile = FileSystemTree.createFile ARCtrl.ArcPathHelper.AssayFileName
        if hasDataMap then
            let dataMapFile = FileSystemTree.createFile ARCtrl.ArcPathHelper.DataMapFileName
            FileSystemTree.createFolder(assayName, [|dataset; protocols; assayFile; readme; dataMapFile|])
        else
            FileSystemTree.createFolder(assayName, [|dataset; protocols; assayFile; readme|])

    static member createStudyFolder(studyName : string, ?hasDataMap) = 
        let hasDataMap = defaultArg hasDataMap false
        let resources = FileSystemTree.createEmptyFolder ARCtrl.ArcPathHelper.StudiesResourcesFolderName
        let protocols = FileSystemTree.createEmptyFolder ARCtrl.ArcPathHelper.StudiesProtocolsFolderName
        let readme = FileSystemTree.createReadmeFile()
        let studyFile = FileSystemTree.createFile ARCtrl.ArcPathHelper.StudyFileName
        if hasDataMap then
            let dataMapFile = FileSystemTree.createFile ARCtrl.ArcPathHelper.DataMapFileName
            FileSystemTree.createFolder(studyName, [|resources; protocols; studyFile; readme; dataMapFile|])
        else
            FileSystemTree.createFolder(studyName, [|resources; protocols; studyFile; readme|])

    static member createWorkflowFolder(workflowName : string, ?hasDataMap) = 
        let hasDataMap = defaultArg hasDataMap false
        let readme = FileSystemTree.createReadmeFile()
        let workflowFile = FileSystemTree.createFile ARCtrl.ArcPathHelper.WorkflowFileName
        if hasDataMap then
            let dataMapFile = FileSystemTree.createFile ARCtrl.ArcPathHelper.DataMapFileName
            FileSystemTree.createFolder(workflowName, [|workflowFile; readme; dataMapFile|])
        else
            FileSystemTree.createFolder(workflowName, [|workflowFile; readme|])

    static member createRunFolder(runName : string, ?hasDataMap) = 
        let hasDataMap = defaultArg hasDataMap false
        let readme = FileSystemTree.createReadmeFile()
        let runFile = FileSystemTree.createFile ARCtrl.ArcPathHelper.RunFileName
        if hasDataMap then
            let dataMapFile = FileSystemTree.createFile ARCtrl.ArcPathHelper.DataMapFileName
            FileSystemTree.createFolder(runName, [|runFile; readme; dataMapFile|])
        else
            FileSystemTree.createFolder(runName, [|runFile; readme|])

    static member createInvestigationFile() = 
        FileSystemTree.createFile ARCtrl.ArcPathHelper.InvestigationFileName

    static member createAssaysFolder(assays : FileSystemTree array) =
        FileSystemTree.createFolder(
            ARCtrl.ArcPathHelper.AssaysFolderName, 
            Array.append [|FileSystemTree.createGitKeepFile()|] assays
        )

    static member createStudiesFolder(studies : FileSystemTree array) =
        FileSystemTree.createFolder(
            ARCtrl.ArcPathHelper.StudiesFolderName,
            Array.append [|FileSystemTree.createGitKeepFile()|] studies
        )

    static member createWorkflowsFolder(workflows : FileSystemTree array) =
        FileSystemTree.createFolder(
            ARCtrl.ArcPathHelper.WorkflowsFolderName, 
            Array.append [|FileSystemTree.createGitKeepFile()|] workflows
        )

    static member createRunsFolder(runs : FileSystemTree array) = 
        FileSystemTree.createFolder(
            ARCtrl.ArcPathHelper.RunsFolderName, 
            Array.append [|FileSystemTree.createGitKeepFile()|] runs
        )